src/lib.rs

```diff
@@ -369,10 +430,29 @@ pub unsafe extern "C" fn path_readlink(
     path_ptr: *const u8,
     path_len: usize,
     buf: *mut u8,
-    _buf_len: Size,
+    buf_len: Size,
     bufused: *mut Size,
 ) -> Errno {
-    unreachable()
+    let fd = wasi_filesystem::Descriptor::from_raw(fd as _);
+
+    let path = slice::from_raw_parts(path_ptr, path_len);
+
+    register_buffer(buf, buf_len);
+
+    let result = fd.readlink_at(path);
+
+    unregister_buffer(buf, buf_len);
+
+    match result {
+        Ok(ref path) => {
+            assert_eq!(path.as_ptr(), buf);
+            assert!(path.len() <= buf_len);
+            *bufused = path.len();
+            forget(path);
```

Comment:
I think this doesn't do anything because `path` was bound with `ref path`. This may be why you were led to add the stub `GlobalAlloc`?

Reply:
Yes, that's it. I put the `forget` in the wrong place. I've now moved it so that it forgets the buffer, and that eliminates the need for the `GlobalAlloc` stub.